pkg/system/rm.go

```diff
@@ -78,3 +78,31 @@ func EnsureRemoveAll(dir string) error {
 		time.Sleep(100 * time.Millisecond)
 	}
 }
+
+// AtomicRemoveAll performs an atomic remove using `EnsureRemoveAll`
+// During removal the passed in dir will not be accessible.
+func AtomicRemoveAll(dir string) error {
+	// best effort to unmount, if this fails (on a mounted fs), the rename/rm will fail below accordingly
+	mount.Unmount(dir)
+
+	renamed := dir + "-removing"
+	err := os.Rename(dir, renamed)
+	switch {
+	case os.IsNotExist(err):
+		// origin dir does not exist, nothing to do
+		return nil
+	case os.IsExist(err):
+		// Some previous remove failed, check if the origin dir exists -- it should not.
+		if _, e := os.Stat(dir); !os.IsNotExist(e) {
+			return errors.Wrap(err, "both rename target and origin dir exist")
```

Comment:
With this code flow, once you'll have `dir+"removed"` entry (say due to someone killing a daemon while removing it, or a race, or some other case), you will never be able to remove either `dir` or `dir+"removed"`.

I have two suggestions:
1. Always try to remove `dir+"-removed"`
2. In case Rename fails (say we have two overmounts on `dir`) , try to remove the `dir` anyway.

Reply:
> Always try to remove `dir+"-removed"`

You can do so before `os.Rename()` (to increase its chances to succeed), maybe with a warning in logs ("Removing stale directory %s" or so).